fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerScanner.java

```diff
@@ -0,0 +1,348 @@
+                            .thenCompose(store -> store.markIndexReadable(common.getIndex()))
+                            .thenApply(ignore -> null)));
+        } else {
+            return buildFuture;
+        }
+    }
+
+    abstract CompletableFuture<Void> scanBuildIndexAsync();
+
+    protected boolean shouldLogBuildProgress() {
+        long interval = common.config.getProgressLogIntervalMillis();
+        long now = System.currentTimeMillis();
+        if (interval == 0 || interval < (now - timeOfLastProgressLogMillis)) {
+            return false;
+        }
+        timeOfLastProgressLogMillis = now;
+        return true;
+    }
+
+    public int getLimit() {
+        // made public to support tests
+        return throttle.getLimit();
+    }
+
+    protected CompletableFuture<Boolean> throttleDelay() {
+        int limit = getLimit();
+        int recordsPerSecond = common.config.getRecordsPerSecond();
+        int toWait = (recordsPerSecond == OnlineIndexerCommon.UNLIMITED) ? 0 : 1000 * limit / recordsPerSecond;
+        return MoreAsyncUtil.delayedFuture(toWait, TimeUnit.MILLISECONDS).thenApply(vignore3 -> true);
+    }
+
+
+    public <R> CompletableFuture<R> buildAsync(@Nonnull BiFunction<FDBRecordStore, AtomicLong, CompletableFuture<R>> buildFunction,
+                                        boolean limitControl,
+                                        @Nullable List<Object> additionalLogMessageKeyValues) {
+
+        return throttle.buildAsync(buildFunction, limitControl, additionalLogMessageKeyValues);
+    }
+
+    private static void timerIncrement(@Nullable FDBStoreTimer timer, FDBStoreTimer.Counts event) {
+        // helper function to reduce complexity
+        if (timer != null) {
+            timer.increment(event);
+        }
+    }
+
+    private static CompletableFuture<Void> updateMaintainerBuilder(SyntheticRecordFromStoredRecordPlan syntheticPlan,
+                                                                   FDBStoredRecord<Message> rec,
+                                                                   IndexMaintainer maintainer,
+                                                                   FDBRecordStore store) {
+        // helper function to reduce complexity
+        if (syntheticPlan == null) {
+            return maintainer.update(null, rec);
+        }
+        // Pipeline size is 1, since not all maintainers are thread-safe.
+        return syntheticPlan.execute(store, rec).forEachAsync(syntheticRecord -> maintainer.update(null, syntheticRecord), 1);
+    }
+
+    protected CompletableFuture<Void> iterateRangeOnly(@Nonnull FDBRecordStore store,
+                                                       @Nonnull RecordCursor<?> cursor,
+                                                       @Nonnull Function<RecordCursorResult<?>, FDBStoredRecord<Message>> getRecord,
+                                                       @Nonnull Consumer<RecordCursorResult<?>> lastResultSet,
```

Comment:
I think this can be something like:

```suggestion
    protected <T> CompletableFuture<Void> iterateRangeOnly(@Nonnull FDBRecordStore store,
                                                       @Nonnull RecordCursor<T> cursor,
                                                       @Nonnull Function<RecordCursorResult<T>, FDBStoredRecord<Message>> getRecord,
                                                       @Nonnull Consumer<RecordCursorResult<T>> lastResultSet,
```

And then some of the cast methods that are currently passed can be made safer. But I haven't actually tried this out